### PR TITLE
fix the incorrect handling of markers in `pip tree`

### PR DESCRIPTION
**Pull request by ChannyClaus**
## Summary
resolves https://github.com/astral-sh/uv/issues/4609

previously, the implementation of `required_with_no_extra` was incorrect, particularly when there are packages that do not require any extras but have other types of markers.

## Test Plan
the existing tests also did cover this (my bad... missed it) but added a smaller test since this bug would've been more obvious with this new test.

### CODE REVIEW

**Comment by ChannyClaus** [comment]
@zanieb also oops noticed the cycle message still uses # 😭 (seems like i messed up the merge conflict)

i can make a follow-up sometime to fix that as well

**Comment by ChannyClaus** [comment]
+ there seem to be a bunch of CI failures from 429...

File: crates/uv/src/commands/pip/tree.rs [hunk 1]
```diff
@@ -82,18 +89,18 @@ pub(crate) fn pip_tree(
...
         .into_iter()
         .filter(|r| {
             r.marker.is_none()
-                || !r
-                    .marker
-                    .as_ref()
-                    .unwrap()
-                    .evaluate_optional_environment(None, &metadata.provides_extras[..])
+                || (r.marker.as_ref().unwrap().evaluate(marker_environment, &[])
+                    && !r.marker.as_ref().unwrap().has_extra())
```
Comment (charliermarsh) [inline crates/uv/src/commands/pip/tree.rs:103]: Is this not sufficient?

```rust
requirement.marker
    .as_ref()
    .map_or(true, |m| m.evaluate(marker_environment, &[]))
```

Why do we need the `has_extra` piece?

Comment (ChannyClaus) [reply crates/uv/src/commands/pip/tree.rs]: oops, you're right, removed 👍

**Comment by charliermarsh** [comment]
Thanks!

**Pull request merged**